apps/opik-backend/src/main/java/com/comet/opik/domain/ChatCompletionService.java

```diff
@@ -16,170 +14,139 @@
+                    request.model());
+        } catch (BadRequestException exception) {
+            handleChunckedException(chunkedOutput, new ErrorMessage(HttpStatus.SC_BAD_REQUEST, exception.getMessage()));
+        }
+
         return chunkedOutput;
     }
 
-    private OpenAiClient getAndConfigureOpenAiClient(ChatCompletionRequest request, String workspaceId) {
-        var llmProvider = getLlmProvider(request.model());
-        var encryptedApiKey = getEncryptedApiKey(workspaceId, llmProvider);
-        return newOpenAiClient(encryptedApiKey);
+    private RetryUtils.RetryPolicy newRetryPolicy() {
+        var retryPolicyBuilder = RetryUtils.retryPolicyBuilder();
+        Optional.ofNullable(llmProviderClientConfig.getMaxAttempts()).ifPresent(retryPolicyBuilder::maxAttempts);
+        Optional.ofNullable(llmProviderClientConfig.getJitterScale()).ifPresent(retryPolicyBuilder::jitterScale);
+        Optional.ofNullable(llmProviderClientConfig.getBackoffExp()).ifPresent(retryPolicyBuilder::backoffExp);
+        return retryPolicyBuilder
+                .delayMillis(llmProviderClientConfig.getDelayMillis())
+                .build();
     }
 
-    /**
-     * The agreed requirement is to resolve the LLM provider and its API key based on the model.
-     * Currently, only OPEN AI is supported, so model param is ignored.
-     * No further validation is needed on the model, as it's just forwarded in the OPEN AI request and will be rejected
-     * if not valid.
-     */
-    private LlmProvider getLlmProvider(String model) {
-        return LlmProvider.OPEN_AI;
+    private <T> Consumer<T> getMessageHandler(ChunkedOutput<String> chunkedOutput) {
+        return item -> {
+            if (chunkedOutput.isClosed()) {
+                log.warn("Output stream is already closed");
+                return;
+            }
+            try {
+                chunkedOutput.write(JsonUtils.writeValueAsString(item));
+            } catch (IOException ioException) {
+                throw new UncheckedIOException(ioException);
+            }
+        };
     }
 
-    /**
-     * Finding API keys isn't paginated at the moment, since only OPEN AI is supported.
-     * Even in the future, the number of supported LLM providers per workspace is going to be very low.
-     */
-    private String getEncryptedApiKey(String workspaceId, LlmProvider llmProvider) {
-        return llmProviderApiKeyService.find(workspaceId).content().stream()
-                .filter(providerApiKey -> llmProvider.equals(providerApiKey.provider()))
-                .findFirst()
-                .orElseThrow(() -> new BadRequestException("API key not configured for LLM provider '%s'".formatted(
-                        llmProvider.getValue())))
-                .apiKey();
+    private Runnable getCloseHandler(ChunkedOutput<String> chunkedOutput) {
+        return () -> {
+            try {
+                chunkedOutput.close();
```

Comment:
Same, if we could I would a void dealing with such details in the service layer

Reply:
Done in https://github.com/comet-ml/opik/pull/957/commits/e0c807683a8bcbe37fae54b48d49a2983ead6390